src/vocab.rs

```diff
@@ -64,24 +71,86 @@ impl PyVocab {
     }
 }
 
-#[pyproto]
-impl PySequenceProtocol for PyVocab {
-    fn __len__(&self) -> PyResult<usize> {
-        let embeds = self.embeddings.borrow();
-        Ok(embeds.vocab().words_len())
-    }
-
-    fn __getitem__(&self, idx: isize) -> PyResult<String> {
+impl PyVocab {
+    fn isize_to_str(&self, mut idx: isize) -> PyResult<String> {
         let embeds = self.embeddings.borrow();
         let words = embeds.vocab().words();
 
+        if idx < 0 {
+            idx += words.len() as isize;
+        }
+
         if idx >= words.len() as isize || idx < 0 {
             Err(exceptions::IndexError::py_err("list index out of range"))
         } else {
             Ok(words[idx as usize].clone())
         }
     }
 
+    fn str_to_indices(&self, query: &str) -> PyResult<WordIndex> {
+        let embeds = self.embeddings.borrow();
+        embeds
+            .vocab()
+            .idx(query)
+            .ok_or_else(|| exceptions::KeyError::py_err(format!("key not found: {}", query)))
+    }
+}
+
+#[pyproto]
+impl PyMappingProtocol for PyVocab {
+    fn __getitem__(&self, query: &PyAny) -> PyResult<PyObject> {
+        let gil = Python::acquire_gil();
+        if let Ok(idx) = query.extract::<isize>() {
+            return self.isize_to_str(idx).map(|s| s.into_py(gil.python()));
+        }
+
+        if let Ok(indices) = query.extract::<Vec<isize>>() {
```

Comment:
Is `foo[[1,2,3]]` or `foo[["x", "y", "z"]]` canonical Python? This seems to overload indexing quite a bit.

Reply:
I dug a bit, turns out that `foo[1,2,3,4]` is canonical and is implemented the same way as `foo[:3]`. This requires extracting indices from `PySlice`, which I'm now implementing. 

Indexing with slices/lists doesn't seem to be a thing with python maps, although I thought it's a nice feature to get all indices for a sentence with a single call.

Reply:
```
>>> foo = [1, 2, 3, 4]
>>> foo[0, 3]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: list indices must be integers or slices, not tuple
```

Reply:
numpy allows you to retrieve items via `foo[[1,2,3,4]]`. I am always a bit disappointed when python doesn't let me index like this into lists.

Reply:
I am also not sure what the purpose of this notation would be (outside some numpy-like cases), no one is going to hardcode indices, right?

The `words[[1,4,5]]` notation is a bit more understandable, because someone might have a list of word indices. But those are served fine by a for-comprehension or `map`?

Reply:
> ```
> >>> foo = [1, 2, 3, 4]
> >>> foo[0, 3]
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> TypeError: list indices must be integers or slices, not tuple
> ```

[0,3] is not a slice, tho. Slices are defined with start, stop and step. `slice(0, 4, 3)` would probably work. `foo[1,2,3]` would also work.

Reply:
> numpy allows you to retrieve items via `foo[[1,8,2,3]]`. I am always a bit disappointed when python doesn't let me index like this into lists.

But `numpy` is really a different case, because looking up indices separately and then converting them back to a dense array is unnecessary expensive. But for lookups that result in references to Python objects, you do not have that penalty.

Reply:
> [0,3] is not a slice, tho. Slices are defined with start, stop and step.

I know what a slice is. I was just repeating your example. You said that `foo[1,2,3,4]` is canonical, but it is not supported by e.g. Python list.

> `slice(0, 4, 3)` would probably work.

Sure.

> `foo[1,2,3]` would also work.

It doesn't. It's not a slice, but a tuple.

Reply:
> But `numpy` is really a different case, because looking up indices separately and then converting them back to a dense array is expensive. But for a series of lookups that result in a Python list of references, you do not have that penalty.

I like the convenience :)

Reply:
I was wrong with that claim, I ran it on the Rust impl and didn't realize I hadn't deleted the `extract::<Vec<isize>>()` arm

Reply:
Also, `vocab[:1000]` would be somewhat useful in sorted vocabularies (though there is no requirement vocabs to be ordered of course). My problem with that is that is actually not a slice that you are getting, but a copy. Unless `pyo3` also permits creating real slices.

Reply:
> I like the convenience :),

Believe me, I can see the convenience. But I have become a firm believer in keeping interfaces small and orthogonal. If you can do it with two simple interfaces (or here: one interface and a language feature), you should prefer that overcomplicating the interface (so that you can do it with one interface).

```python
[vocab[idx] for idx in l]
```

is not that much more work than

```python
vocab[l]
```

plus you don't have to guess the return type ;).

Of course, there are sometimes good exceptions (as you pointed out with `numpy`) and it's not an exact science. E.g. we could also axe the `embedding` method, since you can do lookups on the vocab + storage manually.

Reply:
> Also, `vocab[:n]` would be somewhat useful in sorted vocabularies (though there is no requirement vocabs to be ordered of course). My problem with that is that is actually not a slice that you are getting, but a copy. Unless `pyo3` also permits creating real slices.

That's actually a good point, I didn't consider that everything is a pointer in python and that slicing lists is cheap because of it. I don't think we can replicate that in Rust unless we wrap everything in Rc which is not feasible.

Reply:
> That's actually a good point, I didn't consider that everything is a pointer in python and that slicing lists is cheap because of it. I don't think we can replicate that in Rust unless we wrap everything in Rc which is not feasible.

Maybe something like:

```
struct VocabSlice {
  embeddings: Rc<_>,
  start: usize,
  end: usize,
}
```

and then the sequence/iteration/... protocols could be implemented on that? You could even cheat and make `Vocab` the slice type and by default `start`/`end` would be the whole range.

Reply:
But that could again be easily mirrored in python through list comprehensions, no?

The argument for me was that `vocab[:n]` would allocate a new PyString for every item in the slice and there's no way around that unless we could pass a Rust string ref to Python that behaves like a PyString, which I believe is impossible. 

It wasn't really that the slice is not an actual view of the list. type([1,2,3][:2]) is also `list`, so I think that Python actually stores the refs in a newly allocated list...

Reply:
> The argument for me was that `vocab[:n]` would allocate a new PyString for every item in the slice and there's no way around that unless we could pass a Rust string ref to Python that behaves like a PyString, which I believe is impossible.

No, in what I proposed that would not be the case. If you do

vocab[:1000]

nothing would be copied. It would just be another `Vocab` instance with its `end` index set to `1000`. Conversions to `PyString` would only happen if you iterate over this sliced vocab and/or get individual entries. But that's also true for the original vocab.

But in this proposal `vocab[:1000]` would be a constant operation, whereas in your proposed implementation it's a linear time operation (due to copying).